web/src/engine/src/main/headless/inputProcessor.ts

```diff
@@ -168,26 +168,24 @@ export class InputProcessor {
    */
   private _processKeyEvent(keyEvent: KeyEvent, textStore: TextStore): ProcessorAction {
     const formFactor = keyEvent.device.formFactor;
-    const fromOSK = keyEvent.isSynthetic;
 
-    // The default OSK layout for desktop devices does not include nextlayer info, relying on modifier detection here.
+    // The default OSK layout for desktop devices does not include nextlayer info, relying on
+    // modifier detection here.
     // It's the OSK equivalent to doModifierPress on 'desktop' form factors.
-    if((formFactor == DeviceSpec.FormFactor.Desktop || !this.activeKeyboard || (this.activeKeyboard instanceof JSKeyboard && this.activeKeyboard.usesDesktopLayoutOnDevice(keyEvent.device))) && fromOSK) {
+    if ((formFactor == DeviceSpec.FormFactor.Desktop || !this.activeKeyboard || (this.activeKeyboard instanceof JSKeyboard && this.activeKeyboard.usesDesktopLayoutOnDevice(keyEvent.device))) && keyEvent.isSynthetic) {
       // If it's a desktop OSK style and this triggers a layer change,
       // a modifier key was clicked.  No output expected, so it's safe to instantly exit.
-      if(this.keyboardProcessor.selectLayer(keyEvent)) {
+      if (this.keyboardProcessor.selectLayer(keyEvent)) {
         return new ProcessorAction();
       }
     }
 
-    // Will handle keystroke-based non-layer change modifier & state keys, mapping them through the physical keyboard's version
-    // of state management.  `doModifierPress` must always run.
-    if (this.keyboardProcessor.doModifierPress(keyEvent, textStore, !fromOSK)) {
+    // Will handle keystroke-based non-layer change modifier & state keys, mapping them through
+    // the physical keyboard's version of state management.  `doModifierPress` must always run.
+    if (this.keyboardProcessor.doModifierPress(keyEvent, textStore, !keyEvent.isSynthetic) && !keyEvent.isSynthetic) {
```

Comment:
I guess the same issue applies with multiple conditions in the `if` statement. If the test has a side-effect (`doModifierPress`), then I think it belongs in its own `if` statement, as a general principle? Probably clearer for code clarity would be to do something like this -- this highlights the call to `doModifierPress` having a side-effect:

```suggestion
    const wasModifierPress = this.keyboardProcessor.doModifierPress(keyEvent, textStore, !keyEvent.isSynthetic);
    if (wasModifierPress && !keyEvent.isSynthetic) {
```

Getting into the bikeshed painting again, but I do think that side-effects are easily missed when buried in `if` statements.

Also -- this clarifies what the result of `doModifierPress` is -- not just a success/failure or 'is condition true'.